File: openconjectures/goemans-cost-counterexample/checkers/dgg_goemans_counterexample_replay.rb
#!/usr/bin/env ruby
dem={"t1"=>15,"t2"=>10,"t3"=>15}; dmax=15
e={%w[s t1]=>[10,2],%w[s t2]=>[6,3],%w[s u]=>[24,0],%w[u t3]=>[10,2],%w[u v]=>[14,0],%w[v t1]=>[5,0],%w[v w]=>[9,0],%w[w t2]=>[4,0],%w[w t3]=>[5,0]}
p={"t1"=>[[%w[s t1]], [%w[s u],%w[u v],%w[v t1]]],"t2"=>[[%w[s t2]], [%w[s u],%w[u v],%w[v w],%w[w t2]]],"t3"=>[[%w[s u],%w[u t3]], [%w[s u],%w[u v],%w[v w],%w[w t3]]]}
raise unless e.values.sum(&:first)==87 && e[%w[s t1]][0]+e[%w[s t2]][0]+e[%w[s u]][0]==40
raise unless e.values.each_with_index.sum{|v,i|v[0]*e.values[i][1]}==58
rows=[]
[0,1].repeated_permutation(3).each do |ch|
 load=Hash.new(0); total=0
 ["t1","t2","t3"].each_with_index do |t,i|
  p[t][ch[i]].each{|a|load[a]+=dem[t]}; total+=dem[t]*p[t][ch[i]].sum{|a|e[a][1]}
 end
 bad=load.each_with_object({}){|(a,v),h|h[a]=v-e[a][0]-dmax if v>e[a][0]+dmax}; rows << [ch,total,bad.empty?,bad]
end
raise unless rows.size==8 && rows.count{|r|r[2]}==4 && rows.select{|r|r[2]}.all?{|r|r[1]>=60}
puts "DGG/Goemans finite counterexample Ruby replay PASS"; rows.each{|r|p r}

File: openconjectures/goemans-cost-counterexample/checkers/dgg_goemans_counterexample_replay.py
#!/usr/bin/env python3
from itertools import product
demand={"t1":15,"t2":10,"t3":15}; dmax=15
edges={("s","t1"):(10,2),("s","t2"):(6,3),("s","u"):(24,0),("u","t3"):(10,2),("u","v"):(14,0),("v","t1"):(5,0),("v","w"):(9,0),("w","t2"):(4,0),("w","t3"):(5,0)}
paths={"t1":[(("s","t1"),),(("s","u"),("u","v"),("v","t1"))],"t2":[(("s","t2"),),(("s","u"),("u","v"),("v","w"),("w","t2"))],"t3":[(("s","u"),("u","t3")),(("s","u"),("u","v"),("v","w"),("w","t3"))]}
x={e:v[0] for e,v in edges.items()}; c={e:v[1] for e,v in edges.items()}
assert sum(x.values())==87 and x[("s","t1")]+x[("s","t2")]+x[("s","u")]==40
assert 10+5==15 and 6+4==10 and 10+5==15 and sum(x[e]*c[e] for e in edges)==58
rows=[]
for ch in product((0,1),repeat=3):
 load={e:0 for e in edges}; total=0
 for i,t in enumerate(("t1","t2","t3")):
  p=paths[t][ch[i]]; total+=demand[t]*sum(c[e] for e in p)
  for e in p: load[e]+=demand[t]
 bad={e:load[e]-x[e]-dmax for e in edges if load[e]>x[e]+dmax}
 rows.append((ch,total,not bad,bad))
assert len(rows)==8 and sum(r[2] for r in rows)==4 and all(r[1]>=60 for r in rows if r[2])
print("DGG/Goemans finite counterexample Python replay PASS")
for r in rows: print(r)
